### DISCUSSION TITLE
Is it possible to get a before value / after value of a controlled component onBlur?

### DISCUSSION THREAD

**Discussion by jvgeee** [react-hook-form/react-hook-form#7835] [Q&A] [answered]
Need to send an analytics event that takes a text input's value when it's focussed + the value once blurred (e.g. once they've finished typing) and sends it to our data warehouse.

Is there a clever in-built way to do this? I'm using `useController` for what it's worth. My first thought was to store the initial value in state on focus, but just curious if there's a RHF-handled method for this that doesn't require me to set my own state

```
const [initValue, setInitValue] = useState("")

const handleFocus () => {
    const current = formControls.getValue('myTextInput')
    setInitValue(current)
}

const handleBlur () => {
    const current = formControls.getValue('myTextInput')
    analytics.send({
       beforeValue: initValue,
       afterValue: current
    })
}

return <TextInput onFocus={handleFocus} onBlur={handleBlur} control={control} />
```

**Comment by JamesWong1999** [comment]
I think you want to store your initial value with on focus?

**Comment by Moshyfawn** [comment]
If I understand you correctly, you want default input value and it's value during the `onBlur` to be sent `onBlur`, right? It seems like you want to take the input's default value from `defaultValues` provided to `useForm` as the `beforeValue` and just pass the `field.value` into the `afterValue`.
```tsx
interface InputProps {
  // all form default values
  defaultValues: DefaultValues<FormValues>
  // or maybe just the input's default value from the defaultValues on useForm
  defaultValue: unknown
}

const Input = ({ defaultValues, defaultValue }: InputProps) => {
  const { field } = useController()
  
  const handleBlur = () => {
    analytics.send({
      // defaultValues
      beforeValue: defaultValues.fieldName,
      // or
      beforeValue: defaultValue,
      afterValue: field.value
    })
    field.onBlur()
  }
  
  return <input {...field} onBlur={handleBlur} />
}
```
If you want the input's "default" value to be the one it has `onFocus`, you will have to setup your own state, preferably via `useRef` to avoid re-renders
```jsx
const Input = ({ defaultValues, defaultValue }: InputProps) => {
  const { field } = useController()
  const beforeValue = useRef(field.value)
  
  const handleFocus = () => {
    beforeValue.current = field.value
  }
  
  const handleBlur = () => {
    analytics.send({
      beforeValue,
      afterValue: field.value
    })
    field.onBlur()
  }
  
  return <input {...field} onFocus={handleFocus} onBlur={handleBlur} />
}
```

RHF doesn't expose any analytics helpers, so it's all up to the user to handle whatever info they wanna store and how to store it

**Reply by Moshyfawn** [reply]
You can surely make the logic into a custom hook to reuse in different inputs
```tsx
interface UseInputAnalyticsOptions {
  defaultValue: string
}

function useInputAnalytics ({ defaultValue }: UseInputAnalyticsOptions) {
  const beforeValue = useRef(defaultValue)
  
  const onBeforeValueChange = (value: string) => {
    beforeValue.current = value
  }
  
  const sendAnalytics = () => {
    analytics.send({
      beforeValue,
      afterValue: field.value
    })
  }
  
  return {
    onBeforeValueChange,
    sendAnalytics
  }
}
```

**Comment by jvgeee** [comment, marked as answer]
Hey guys, no, I don't want just the default value of the input, I want the value onFocus and onBlur (e.g. if it's got text in it, and then you click in and type extra text, I want the before/after value).

Seems like there's nothing in RHF for this and it's just best handled via component state. Happy with that! Cheers for the help.
